### ISSUE TITLE
Significant Discrepancy in Results on ScanNet Dataset with Provided Settings

### ISSUE THREAD

**Issue by tj12323** [sharinka0715/semantic-gaussians#28]
Thank you for your insightful work and for sharing the code for "Semantic Gaussians: Open-Vocabulary Scene Understanding with 3D Gaussian Splatting". I am attempting to replicate the results described in your paper using the provided settings and dataset details. However, I am observing significant discrepancies in the results compared to those reported in the paper. Below are the details of my setup and results:

### Setup
- **Dataset**: ScanNet  
  - Scenes used: As mentioned in the supplementary file of the paper ("Table 4. Scenes used for evaluations in our experiments").  
- **Configuration Files**:
  - `official_train.yaml`, `fusion_scannet.yaml`, and `eval.yaml` were used as per the repository. My configuration details are attached for your reference:
    - **Train Configuration**:(official_train.yaml)  
      ```yaml   
      scene:
        scene_path: "dataset_link/ScanNet/scene0050_02"
        test_cameras: True
        colmap_images: "images"
        colmap_eval_hold: 8
        downscale_ratio: 1
        white_background: False
        device: "cuda:0"
        
      pipeline:
        convert_shs_python: False
        compute_cov3d_python: False
        debug: False
        seed: 0
      
      model:
        sh_degree: 3
        model_dir: ~
        load_iteration: -1
        device: "cuda:0"
        random_init: False
      
      train:
        exp_name: "scene0050_02"
        iterations: 30000
        num_workers: 8
        test_iterations: [100, 7000, 30000]
        save_iterations: [7000, 30000]
        checkpoint_iterations: []
        cut_edge: True  # If ScanNet, set True to cut the 1% black edge
      
        position_lr_init: 0.00016
        position_lr_final: 0.0000016
        position_lr_delay_mult: 0.01
        position_lr_max_steps: 10000
        feature_lr: 0.0025
        opacity_lr: 0.05
        scaling_lr: 0.005
        rotation_lr: 0.001
        percent_dense: 0.01
        lambda_dssim: 0.2
        densification_interval: 100 # 1000 # 100 # 2000
        opacity_reset_interval: 3000 #2000 # 3000 # 2000
        densify_from_iter: 500
        densify_until_iter: 15000 # 10000 # 15000 #10000
        densify_grad_threshold: 0.0002
        random_background: False
      ```
    - **Fusion Configuration**:(fusion_scannet.yaml)  
      ```yaml
      scene:
        scene_path: "dataset_link/ScanNet"
        dataset_name: "cocomap"
        test_cameras: False
        colmap_images: "images"
        colmap_eval_hold: 8
        downscale_ratio: 1
        white_background: False
        device: "cuda:0"
        
      pipeline:
        convert_shs_python: False
        compute_cov3d_python: False
        debug: False
        seed: 0
      
      model:
        sh_degree: 3
        model_dir: "output"
        dynamic: False
        load_iteration: -1
        device: "cuda:0"
      
      fusion:
        img_dim: [648, 484]
        num_workers: 8
        model_2d: lseg # choose from openseg, lseg, samclip, vlpart
        depth: render # choose from image, render, surface, none
        depth_scale: 1000.0
        visibility_threshold: 0.05
        cut_boundary: 10
        n_split_points: 80000 # train: 80000, eval: 999999999 (large enough)
        num_rand_file_per_scene: 5 # train: 5, eval: 1
        out_dir: "output_fusion"
      ```
    - **Eval Configuration**:(eval.yaml)  
      ```yaml
      scene:
        scene_path: "dataset_link/ScanNet"
        dataset_name: "cocomap"
        test_cameras: False
        colmap_images: "images"
        colmap_eval_hold: 8
        downscale_ratio: 1
        white_background: False
        device: "cuda:0"
        
      pipeline:
        convert_shs_python: False
        compute_cov3d_python: False
        debug: False
        seed: 0
      
      model:
        sh_degree: 3
        model_dir: "output"
        load_iteration: -1
        device: "cuda:0"
      
      fusion:
        out_dir: "output_fusion"
        model_2d: lseg
      
      distill:
        model_3d: MinkUNet34A
        model_dir: "/PATH/TO/YOUR/OWN"
        text_model: openseg
        voxel_size: 0.02
        iteration: -1
        feature_type: all
      
      eval:
        eval_mode: 2d # choose from 2d, 3d, 2d_and_3d, pretrained, labelmap
        width: 648
        height: 484
        pred_on_3d: True
        feature_fusion: concat # choose from concat, argmax
      ```
### Observed Results
Here are the evaluation metrics I obtained:
```plaintext
92% 11/12 [1:24:32<09:10, 550.95s/it]Found pose directory, assuming ScanNet data set!
Loading Training Cameras...
Loading Test Cameras...
100% 12/12 [1:29:56<00:00, 449.67s/it]
evaluating 5314399013 points...
num_classes: 20
classes          IoU
----------------------------
wall          : 0.719   (1295528485/1801455319)
floor         : 0.762   (1133327540/1486641475)
cabinet       : 0.648   (129547097/199810560)
bed           : 0.885   (215846026/244001756)
chair         : 0.700   (636776094/909563529)
sofa          : 0.644   (142067918/220452064)
table         : 0.583   (301940732/518184555)
door          : 0.558   (218474322/391527868)
window        : 0.324   (44097680/136152146)
shelves       : 0.505   (78769796/156005523)
counter error!
curtain       : 0.506   (32080368/63347402)
ceiling       : 0.541   (8371694/15461714)
refridgerator : 0.005   (103581/20526124)
television    : 0.267   (8415268/31561356)
person error!
toilet        : 0.743   (32352504/43521921)
sink          : 0.312   (5221984/16754307)
lamp          : 0.036   (449910/12592755)
bag           : 0.229   (5365109/23473713)
Mean IoU 0.4981862322861649
Mean Acc 0.6631818747336289
```

### Expected Results
From the paper, I expected a Mean IoU and Mean Accuracy much closer to the reported results, which are significantly higher.

### Request
Could you kindly provide clarification on:
1. Whether there are any additional implementation details or preprocessing steps not explicitly stated in the paper or the repository?  
2. Any potential discrepancies or nuances in the training or evaluation pipeline that I might have missed?  
3. Recommendations for debugging this issue or improving the results?

Your guidance would be immensely helpful. Thank you for your time and support!

**Comment by sharinka0715** [comment]
Hello,

Thank you very much for your support and interest in our work. We highly value your feedback regarding the experimental results. After reviewing your issue, we believe there might be some issues with your experimental setup, which could be the reason why you were unable to reproduce our results.

### About Settings

Differences in experimental settings can lead to varying degrees of discrepancy in results. However, in our view, **a critical setting that could significantly affect the outcome is the `n_split_points` and `num_rand_file_per_scene` parameters in the `fusion_scannet.yaml` file**. If you aim to evaluate the results of 2D fusion, these parameters should be set to `999999999` (large enough) and `1`, respectively, instead of the current settings of `80000` and `5`. The current configuration results in the generation of 5 `.pt` files per scene in the fusion directory, each containing only `80,000` Gaussian points. This setting is intended for data augmentation in training the 3D distillation network and should not be activated when evaluating the 2D fusion results alone. We apologize for not clarifying this point in our paper.

Additionally, the number of training iterations for 3DGS used in the paper is `10000`, not `30000`, and some intervals have been adjusted accordingly. This was mentioned in the paper but was not correctly reflected in the code. Other aspects should not differ significantly.

Furthermore, we recommend trying the `eval_segmentation.py` script with `eval.eval_mode` set to `pretrained` and `fusion.model_2d` set to `lseg`. In this configuration, the model will directly evaluate the segmentation capability of the 2D segmentation foundation model without utilizing 3DGS. In any case, our segmentation results (and those of other baselines) should not underperform compared to the 2D foundation model.

We hope that by adjusting these settings, you will achieve more satisfactory experimental results. 

### About Our Results

Additionally, to minimize further communication overhead, we have open-sourced all the data for the 12 scenes used in our experiments, including:

1. The preprocessed ScanNet dataset `(scannet_half)`, where the images have been resized to `(648, 484)`, and every 5th frame is sampled from the original trajectory.
2. The 3DGS model trained for 10,000 steps along with the corresponding configuration files `(scannet_gaussians)`.
3. The `.pt` files obtained through 2D fusion `(fusion_lseg)`.

By correctly loading these three components into the eval.yaml file, you should be able to reproduce the same results as we did. Download link: https://drive.google.com/file/d/1YQZMid8rAkmf5rxRJqzfhVf2AgjJr4Vp/view?usp=drive_link

Here are the experimental results we obtained on these 12 scenes. In the experiments described in the paper, to speed up the process, we uncommented lines 222 and 223, evaluating one out of every five images to accelerate the evaluation process. Even if every image were evaluated, the results would not differ significantly.

```
classes          IoU
----------------------------
wall          : 0.872   (53991965/61882441)
floor         : 0.878   (52047721/59263823)
cabinet       : 0.732   (5354857/7318175)
bed           : 0.943   (9258700/9822181)
chair         : 0.747   (26497377/35448673)
sofa          : 0.674   (6421762/9522712)
table         : 0.714   (14810536/20754617)
door          : 0.610   (9287895/15221339)
window        : 0.349   (1931892/5534900)
shelves       : 0.805   (4901923/6092521)
counter error!
curtain       : 0.546   (1373292/2517159)
ceiling       : 0.720   (453651/629800)
refridgerator : 0.000   (     0/781241)
television    : 0.365   (396092/1084170)
person error!
toilet        : 0.953   (1361865/1428468)
sink          : 0.354   (225286/636029)
lamp          : 0.391   (218902/559193)
bag           : 0.321   (270257/842825)
Mean IoU 0.6097367666637112
Mean Acc 0.7659856329463729
```

**Comment by tj12323** [comment]
Thank you so much for your prompt and detailed response. I greatly appreciate your clarification on the critical settings and the additional resources you have provided.I look forward to seeing the updated results and will keep you posted on my progress. Thanks again for your invaluable help!
